Profiling: add a retry logic

r1viollet opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

#### Retry Logic Implementation:

Add retry mechanism for profile uploads with exponential backoff (3 attempts, 100ms base delay, 2s max delay)
Clone requests to enable proper retry handling while preserving original context timeout
Implement fresh transport creation for retry attempts to avoid stale connection issues

**Connection Pool Issue Mitigation:**
Address the "poisoned connection" problem where:
HTTP client maintains persistent connections to load balancer
Backend failure causes load balancer's backend server to become unhealthy
Client reuses existing connection to load balancer
Load balancer returns HTTP 502 Bad Gateway but keeps connection open
Subsequent requests reuse the same "poisoned" connection → repeated 502s
Solution: Use transport.Clone() for retry attempts to force fresh connections and bypass stale connection pools.

#### Enhanced Error Classification:

Test and classify retryable vs non-retryable errors for both connection-level and HTTP status code errors
Implement proper HTTP status code semantics:
408 Request Timeout: Body read timeouts (more accurate than 502)
503 Service Unavailable: Retryable connection errors
504 Gateway Timeout: Context deadline exceeded
502 Bad Gateway: Non-retryable errors (default)

#### Improved Observability:

Add comprehensive logging around request body read failures with payload size and duration metrics
Enhance retry attempt logging with detailed error context and fresh transport usage
Increase log throttling limits (50/second) for better debugging visibility during issues
Add WARN-level logging for all retry attempts and connection errors for production debugging

### Motivation

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

r1viollet committed:
Profiling: add a retry logic

- Test errors which can be retried (error codes and HTTP error codes)
- Clone the request to allow retries
- Add a settings on retry with backoff logics

r1viollet committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into r1viollet/retry_profile_upload

ajgajg1134 commented on pkg/trace/api/profiles.go:
```diff
@@ -157,12 +164,148 @@ type multiTransport struct {
+	}
+
+	// Check for URL errors that might be retryable
+	if urlErr, ok := err.(*url.Error); ok {
```
I think you can drop this check, as far as I can tell a url.Error should never be temporary as I believe these can only happen from failing to parse the provided URL

ajgajg1134 commented on pkg/trace/api/profiles.go:
```diff
@@ -173,27 +316,30 @@ func (m *multiTransport) RoundTrip(req *http.Request) (rresp *http.Response, rer
 	}
+
+	// Multiple targets - read body once for all requests
 	slurp, err := io.ReadAll(req.Body)
```
It seems like the retry function here is duplicating the reading of the body here?

r1viollet commented on pkg/trace/api/profiles.go:
```diff
@@ -157,12 +164,148 @@ type multiTransport struct {
+		// Read the request body once for all retry attempts
+		if req.Body != nil {
+			var err error
+			bodyBytes, err = io.ReadAll(req.Body)
```
so reading the body is not great, though it could get consumed by the initial request even if we need to retry

r1viollet committed:
Profiling - Retry logic adjustements

- Avoid the double read of the request body
- Minor adjustement on error management (using error.As)

ajgajg1134 committed:
fix linter errors

r1viollet committed:
Profiling: Re-create a fresh transport object on retry

Forces new connections

r1viollet committed:
Profiler - Add better log diagnostics

We are investigating 502 errors.
The aim is to have a better understanding of the size and timings causing the issues.

r1viollet committed:
Profiling upload - send back more precise errors to profiling libraries

- Ensure we return status codes that are more accurate
- Add some debug logs to explain the errors

r1viollet committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into r1viollet/retry_profile_upload

r1viollet committed:
network_test tool

The aim is to create different network failures.
This was useful to reproduce system pressure when attempting to upload profiles.

r1viollet committed:
profiling - minor adjustement on log throttling

Bring the throttle down to 50 to avoid issues in staging

r1viollet commented:
Closing this one in favour of https://github.com/DataDog/datadog-agent/pull/40045
No strong evidence that agent side retries are needed.
